File: codes/rust/503.下一个更大元素-ii.rs
/*
 * @lc app=leetcode.cn id=503 lang=rust
 *
 * [503] 下一个更大元素 II
 */

// @lc code=start
impl Solution {
    pub fn next_greater_elements(nums: Vec<i32>) -> Vec<i32> {
        let mut result = vec![-1; nums.len() * 2];
        let mut stack = Vec::new();
        let double = nums.repeat(2);
        for (idx, v) in double.iter().enumerate() {
            while !stack.is_empty() && double[*stack.last().unwrap()] < *v {
                let last_idx = stack.pop().unwrap();
                result[last_idx] = *v;
            }
            stack.push(idx);
        }
        result.into_iter().take(nums.len()).collect()
    }
}
// @lc code=end

File: codes/golang/503.下一个更大元素-ii.go

func nextGreaterElements(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	for i := 0; i < n; i++ {
		res[i] = -1
	}
	stack := make([]int, 0)
	for i := 0; i < n*2; i++ {
		for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[i%n] {
			lastIdx := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			res[lastIdx] = nums[i%n]
		}
		stack = append(stack, i%n)
	}
	return res
}